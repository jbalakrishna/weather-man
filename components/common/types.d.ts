export type Error = {
  error: string;
  message: string;
  errorType: number;
};
