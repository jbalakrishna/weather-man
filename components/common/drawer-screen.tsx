import {
  DrawerContentScrollView,
  DrawerItemList,
} from "@react-navigation/drawer";
import { Switch, Text, View } from "react-native";
import useSettingsStore from "./settings/store";

type TSwitchItemProps = {
  handleSwitchToggle: () => void;
  enabled: boolean;
  text: string;
};

const SwitchItem = ({
  handleSwitchToggle,
  enabled,
  text,
}: TSwitchItemProps) => {
  return (
    <View className="flex-1 flex-row align-center justify-between pr-16 border-b pb-2 border-slate-200">
      <Text className="font-medium text-l text-slate-600 py-2">{text}</Text>
      <Switch
        value={enabled}
        trackColor={{ false: "#767577", true: "#81b0ff" }}
        thumbColor="#f4f3f4"
        onValueChange={handleSwitchToggle}
      ></Switch>
    </View>
  );
};

const DrawerScreen = (props: any) => {
  const settingStore = useSettingsStore();
  const handleUnitsChange = () => {
    settingStore.setUnits(
      settingStore.units === "metric" ? "imperial" : "metric"
    );
  };

  const handleOfflineModeChange = () => {
    settingStore.setOfflineMode(!settingStore.offlineMode);
  };

  return (
    <View className="flex-1 py-12 bg-slate-100">
      <DrawerContentScrollView {...props}>
        <View className="flex-1 gap-12 ">
          <View className="flex-1 bg-slate-100">
            <DrawerItemList {...props} />
          </View>
          <View className="flex-1 px-4 pt-4 gap-4 bg-slate-100 border-t-2 border-slate-200 ">
            <Text className="font-bold text-xl text-slate-600">
              Preferences
            </Text>
            <SwitchItem
              handleSwitchToggle={handleUnitsChange}
              enabled={settingStore.units === "imperial"}
              text={"Use Imperial Units\n(Fahrenheit, mph)"}
            />
            <SwitchItem
              handleSwitchToggle={handleOfflineModeChange}
              enabled={settingStore.offlineMode}
              text={"Offline mode"}
            />
          </View>
        </View>
      </DrawerContentScrollView>
    </View>
  );
};

export default DrawerScreen;
