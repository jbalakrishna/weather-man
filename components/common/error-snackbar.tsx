import useIsOffline from "@/components/common/hooks/useIsOffline";
import { delay } from "lodash";
import { useEffect, useRef } from "react";
import { Animated, Text } from "react-native";
const errorContainerClass = "bg-red-600 px-8 justify-center align-center";
const offlineContainerClass = "bg-slate-200 px-8 justify-center align-center";

const ErrorSnackbar = ({ error }: any) => {
  const height = useRef(new Animated.Value(0)).current;
  const showOfflineMode = useIsOffline();

  useEffect(() => {
    if (error || showOfflineMode) {
      Animated.timing(height, {
        toValue: 44,
        duration: 200,
        useNativeDriver: false,
      }).start();
      if (error) {
        delay(() => {
          Animated.timing(height, {
            toValue: 0,
            duration: 200,
            useNativeDriver: false,
          }).start(() => {
            height.setValue(0);
          });
        }, 5000);
      }
      return;
    }

    if (!showOfflineMode) {
      Animated.timing(height, {
        toValue: 0,
        duration: 200,
        useNativeDriver: false,
      }).start();
    }

    return () => {
      height.setValue(0);
    };
  }, [error, showOfflineMode]);

  return (
    <Animated.View
      className={error ? errorContainerClass : offlineContainerClass}
      style={{ height }}
    >
      {error ? (
        <Text className="text-slate-100 font-semibold">
          {
            "There was a problem with your last request. \nPlease try again after some time."
          }
        </Text>
      ) : (
        <Text className="text-slate-500 font-semibold">
          {"Looks like you're offline!"}
        </Text>
      )}
    </Animated.View>
  );
};

export default ErrorSnackbar;
