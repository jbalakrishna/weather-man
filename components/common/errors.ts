import { Error } from "./types";

export enum ErrorType {
  LOCATION_DENIED = 1,
  LOCATION_GENERIC = 2,
  GENERIC = 3,
  WEATHER_FETCH_ERROR = 4,
}

export const WeatherFetchError: Error = {
  error: "Weather fetch error",
  message: "Unable to fetch weather now, please try again later.",
  errorType: ErrorType.WEATHER_FETCH_ERROR,
};

export const LocationErrorDenied: Error = {
  error: "denied",
  errorType: ErrorType.LOCATION_DENIED,
  message:
    "Permission was denied, please try again or open settings to allow the app to access your location.",
};

export const LocationErrorGeneric: Error = {
  error: "generic",
  errorType: ErrorType.LOCATION_GENERIC,
  message: "Unable to fetch location now, please try again later.",
};
