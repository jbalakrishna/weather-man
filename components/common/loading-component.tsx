import { Fragment, useEffect, useRef } from "react";
import { ActivityIndicator, Animated, Text, View } from "react-native";

type TLoadingScreenProps = {
  visible: boolean;
};

const LoadingScreen = (props: TLoadingScreenProps) => {
  const opacity = useRef(new Animated.Value(0)).current;

  useEffect(() => {
    if (props.visible) {
      Animated.timing(opacity, {
        toValue: 1,
        duration: 150,
        useNativeDriver: false,
      }).start();
    }
  }, [props.visible]);

  return props.visible ? (
    <Animated.View
      className="absolute top-0 bottom-0 left-0 right-0 bg-slate-100 opacity-75 items-center justify-center"
      style={{ opacity }}
    >
      <View className="flex-1 gap-4 justify-center">
        <ActivityIndicator size="large" color="grey" />
        <Text className="text-slate-500 font-semibold">
          Fetching your weather info...
        </Text>
      </View>
    </Animated.View>
  ) : (
    <Fragment></Fragment>
  );
};

export default LoadingScreen;
