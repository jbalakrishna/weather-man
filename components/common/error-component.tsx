import Ionicons from "@expo/vector-icons/Ionicons";
import { Fragment, useEffect, useRef } from "react";
import { Animated, Image, Text, TouchableOpacity, View } from "react-native";
import { ErrorType } from "./errors";
import { Error } from "./types";

type TErrorScreenProps = {
  error: Error | null;
  handleRetry: () => void;
};

const errorImage = "https://i.ibb.co/1vHrDq1/1326461-2.png";

const ErrorScreen = ({ error, handleRetry }: TErrorScreenProps) => {
  const opacity = useRef(new Animated.Value(0)).current;

  useEffect(() => {
    if (error) {
      Animated.timing(opacity, {
        toValue: 1,
        duration: 150,
        useNativeDriver: false,
      }).start();
    }
  }, [error]);

  const canRetry =
    error?.errorType === ErrorType.GENERIC ||
    error?.errorType === ErrorType.WEATHER_FETCH_ERROR;

  return error ? (
    <Animated.View
      className="flex-1 bg-slate-100 opacity-75 items-center justify-center"
      style={{ opacity }}
    >
      <View className="flex-1 gap-4 justify-center align-center">
        <Image
          resizeMode="contain"
          className="h-40"
          source={{ uri: errorImage }}
        />

        <Text className="px-16 text-center text-red-500 font-semibold">
          {error.message}
        </Text>
        {canRetry && (
          <TouchableOpacity onPress={handleRetry}>
            <View className="justify-center py-2 px-6 flex-row gap-2">
              <Text className="text-right text-base font-semibold font-xl text-slate-500">
                Retry
              </Text>
              <Ionicons name="refresh-sharp" size={18} />
            </View>
          </TouchableOpacity>
        )}
      </View>
    </Animated.View>
  ) : (
    <Fragment></Fragment>
  );
};

export default ErrorScreen;
