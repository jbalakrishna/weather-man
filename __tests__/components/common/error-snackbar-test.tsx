import ErrorSnackbar from "@/components/common/error-snackbar";
import { ErrorType } from "@/components/common/errors";
import { act, render, waitFor } from "@testing-library/react-native";

jest.mock("@/components/common/hooks/useIsOffline", () => () => true);

describe("ErrorSnackbar 1", () => {
  test("error mode should render correctly", async () => {
    const promise = Promise.resolve();
    const tree = render(
      <ErrorSnackbar
        error={{
          errorType: ErrorType.GENERIC,
          message: "any msg",
        }}
      />
    ).toJSON();

    await waitFor(() => {
      expect(tree).toMatchSnapshot();
    });
    await act(async () => {
      await promise;
    });
  });
});

describe("ErrorSnackbar 2", () => {
  test("offline mode should correctly", async () => {
    const promise = Promise.resolve();
    const tree = render(<ErrorSnackbar error={null} />).toJSON();

    await waitFor(() => {
      expect(tree).toMatchSnapshot();
    });
    await act(async () => {
      await promise;
    });
  });
});
