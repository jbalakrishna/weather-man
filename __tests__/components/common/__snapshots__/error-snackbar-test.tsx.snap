// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorSnackbar 1 error mode should render correctly 1`] = `
<View
  className="bg-red-600 px-8 justify-center align-center"
  collapsable={false}
  style={
    {
      "height": 0,
    }
  }
>
  <Text
    className="text-slate-100 font-semibold"
  >
    There was a problem with your last request. 
Please try again after some time.
  </Text>
</View>
`;

exports[`ErrorSnackbar 2 offline mode should correctly 1`] = `
<View
  className="bg-slate-200 px-8 justify-center align-center"
  collapsable={false}
  style={
    {
      "height": 0,
    }
  }
>
  <Text
    className="text-slate-500 font-semibold"
  >
    Looks like you're offline!
You can still explore but information maybe outdated
  </Text>
</View>
`;
